src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MeterListener.cs

```diff
@@ -0,0 +1,243 @@
+    /// </summary>
+#if ALLOW_PARTIALLY_TRUSTED_CALLERS
+        [System.Security.SecuritySafeCriticalAttribute]
+#endif
+    public sealed class MeterListener : IDisposable
+    {
+        // We use LikedList here so we don't have to take any lock while iterating over the list as we always hold on a node which be either valid or null.
+        // LinkedList is thread safe for Add, Remove, and Clear operations.
+        private static LinkedList<MeterListener> s_allStartedListeners = new LinkedList<MeterListener>();
+
+        // List of the instruments which the current listener is listening to.
+        private LinkedList<Instrument> _enabledMeasurementInstruments = new LinkedList<Instrument>();
+        private bool _disposed;
+
+        // We initialize all measurement callback with no-op operations so we'll avoid the null checks during the execution;
+        private MeasurementCallback<byte>    _byteMeasurementCallback    = (instrument, measurement, tags, state) => { /* no-op */ };
+        private MeasurementCallback<short>   _shortMeasurementCallback   = (instrument, measurement, tags, state) => { /* no-op */ };
+        private MeasurementCallback<int>     _intMeasurementCallback     = (instrument, measurement, tags, state) => { /* no-op */ };
+        private MeasurementCallback<long>    _longMeasurementCallback    = (instrument, measurement, tags, state) => { /* no-op */ };
+        private MeasurementCallback<float>   _floatMeasurementCallback   = (instrument, measurement, tags, state) => { /* no-op */ };
+        private MeasurementCallback<double>  _doubleMeasurementCallback  = (instrument, measurement, tags, state) => { /* no-op */ };
+        private MeasurementCallback<decimal> _decimalMeasurementCallback = (instrument, measurement, tags, state) => { /* no-op */ };
+
+        /// <summary>
+        /// Creates a MeterListener object.
+        /// </summary>
+        public MeterListener() { }
+
+        /// <summary>
+        /// Callbacks to get notification when an instrument is published.
+        /// </summary>
+        public Action<Instrument, MeterListener>? InstrumentPublished { get; set; }
+
+        /// <summary>
+        /// Callbacks to get notification when stopping the measurement on some instrument.
+        /// This can happen when the Meter or the Listener is disposed or calling <see cref="Dispose" /> on the listener.
+        /// </summary>
+        public Action<Instrument, object?>? MeasurementsCompleted { get; set; }
+
+        /// <summary>
+        /// Start listening to a specific instrument measurement recording.
+        /// </summary>
+        /// <param name="instrument">The instrument to listen to.</param>
+        /// <param name="state">A state object which will be passed back to the callback getting measurements events.</param>
+        public void EnableMeasurementEvents(Instrument instrument, object? state = null)
+        {
+            if (instrument is null || _disposed)
```

Comment:
I don't think we need to hold this PR on it, but I believe we'll need to figure out what concurrency guarantees we are planning to give. For example it feels perfectly reasonable for code component A to be invoking Meter.Dispose() concurrently with code component B that is invoking Listener.EnableMeasurementEvents(). If that happened in the current implementation the listener may never receive the MeasurementsComplete callback. A similar issue could happen Listener.Dispose() was called concurrently with Listener.EnableMeasurementEvents(). Some other races lead to double notifications rather than missing ones.

In general I think we need to resolve:
1. Are listener publish/complete callbacks serialized or do listeners need to handle multiple concurrent callbacks?
2. Are listener callbacks guaranteed to occur exactly once or are we allowed to give duplicate callbacks?
3. Which methods are legal to call concurrently?

Reply:
I'll look at that after we merge this PR.

Reply:
This has been addressed by the PR https://github.com/dotnet/runtime/pull/53204